plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -74,92 +88,252 @@ The current value of the underlying asset is determined by an oracle. See note
 
 -}
 
+-- | Run the futures contract. Note that this does not make the initial 
+--   transaction (see 'writeInitialTx')
+runContract
+    :: ContractActions r
+    => Future
+    -> PubKey
+    -> PubKey
+    -> Contract r ()
+runContract f long short =
+    flip runReaderT f $ do
+        fd <- initialData long short
+        addr <- contractAddress
+        let initMp  = AM.addAddress addr Data.Monoid.mempty
+            st = FutureState { knownOutputs = initMp, futureData = fd }
+        flip evalStateT st $ 
+            (liftBase (endpoint "start") >>= futureContract)
+            <|>
+            (liftBase (endpoint @() "initialise") >> writeInitialTx)
+
+{- Note [Initialising the futures contract]
+
+The margin accounts of both parties (long & short) are kept in the same 
+transaction output. To initialise the contract we need a transaction that
+locks the initial margins of both parties with the contract script, as is done 
+in 'writeInitialTx', which is called from the "initialise" endpoint of the 
+contract. In the test suite we simply call "initialise" from one of the 
+emulator's wallets. To make this more realistic we could use an escrow contract 
+that takes the initial margins and then locks them using the script.
+
+TODO: Implement escrow contract that pays to a script address, and use it here.
+
+-}
+
+data Role = Long | Short
+    deriving (Generic, Show, FromJSON, ToJSON)
+
+-- | The current state of the futures contract.
+data FutureState =
+    FutureState
+        { knownOutputs :: AddressMap
+        -- ^ The unspent outputs (usually exactly 1) of the contract instance.
+        , futureData   :: FutureData
+        -- ^ The current margin accounts
+        }
+
+type MonadFuture r m =
+    ( MonadState FutureState m
+    , MonadReader Future m
+    , MonadBase (Resumable (Eff r)) m)
+-- TODO: ResumableT ?
+
+-- | Spend all known outputs at the contract address using the 'FutureRedeemer'.
+--   Use 'redeemWith0' instead if the transaction does not have a pay-to-script 
+--   output.
+redeemWith
```

Comment:
I'd maybe call these `redeemAllWith`, I kept forgetting the looked for all the outputs.